Problem: Race condition on token get/create

steve-gregory opened the pull request:
Solution: Fix the atomic transaction behavior.

If an integrity error is thrown, the reason will be that the object was already created.
In these instances, go ahead and re-attempt the atomic transaction.
Update django-cyverse-auth to 1.0.16 -- Fix atomic transaction block

Fixes #10 .

steve-gregory committed:
Problem: Race condition on token get/create

Solution: Fix the atomic transaction behavior &&  update django-cyverse-auth to 1.0.16

Backstory: How did the IntegrityError occur?
 - When the application receives multiple authenticated requests to DRF, we enter a race condition:
  - Call 1 and Call 2 received instantaneously and handled by separate python processes
  - Call 1 and Call 2 are valid, so both will pass the Authenticate() methods.
  - During authenticate(), a valid Token will be stored in the database. This is where the race condition begins:
    - Both calls `ask DB for Token` at the same time.
    - Both calls `create Token in DB` at the same time.
    - The first call to `create` will succeed and the second call to `create` will throw an IntegrityError.

Backstory: Previous attempts to handle the IntegrityError
 Previous attempts included `transaction.atomic` but attempted to handle the IntegrityError _inside the atomic block_ which caused:
    `TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.`

The latest solution uses update_or_create (Thanks to @cdosborn for mentioning this) and removes `get_or_create_user_and_token()` in favor of a more straightforward `get_or_create_user()`, `get_or_create_token()`
Additionally, `get_or_create_token()` will now expect `user` and not `username`.

cdosborn commented on django_cyverse_auth/models.py:
```diff
@@ -150,72 +150,85 @@ def create_access_token(token_key, token_expire, issuer):
+    if email and user.email != email:
         changed = True
-        user.email = attributes['email']
+        user.email = email
```
This whole pattern is solved by `update_or_create`

cdosborn commented on django_cyverse_auth/models.py:
```diff
@@ -150,71 +151,70 @@ def create_access_token(token_key, token_expire, issuer):
+        auth_token = get_or_create_token(user_profile['username'], token_key, token_expire, remote_ip, issuer)
+    except IntegrityError:
+        auth_token = get_or_create_token(user_profile['username'], token_key, token_expire, remote_ip, issuer)
+
```
The method body can be replaced by:
`return get_or_create_token(user_profile['username'], token_key, token_expire, remote_ip, issuer)`

The better fix is to nuke this method and replace all occurences of `get_or_create_user_and_token` with `get_or_create_token`

cdosborn commented on django_cyverse_auth/models.py:
```diff
@@ -150,71 +151,70 @@ def create_access_token(token_key, token_expire, issuer):
     (And token_key, expiration, issuer.. If available)
     """
-    User = get_user_model()
+    token_expire = auth_user_token.update_expiration(token_expire)
```
auth_user_token is undefined

cdosborn commented on django_cyverse_auth/models.py:
```diff
@@ -150,71 +151,70 @@ def create_access_token(token_key, token_expire, issuer):
+    logger.debug(
+        "%s token - %s" ,
+        "Created" if created else "Retrieved",
+        token_key)
```
👍

steve-gregory committed:
Update version to 1.1.0

- removed `get_or_create_user_and_token`
- changed arguments for `get_or_create_token`
- Fixed race conditions on `get_or_create_user`
- Fixed race conditions on `get_or_create_token`
